src/disco/gui/fd_gui_printf.c

```diff
@@ -0,0 +1,1141 @@
+  jsonp_strip_trailing_comma( gui );
+  fd_http_server_printf( gui->http, "}," );
+}
+
+static void
+jsonp_open_array( fd_gui_t *   gui,
+                  char const * key ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":[", key );
+  else                   fd_http_server_printf( gui->http, "[" );
+}
+
+static void
+jsonp_close_array( fd_gui_t * gui ) {
+  jsonp_strip_trailing_comma( gui );
+  fd_http_server_printf( gui->http, "]," );
+}
+
+static void
+jsonp_ulong( fd_gui_t *   gui,
+             char const * key,
+             ulong        value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%lu,", key, value );
+  else                   fd_http_server_printf( gui->http, "%lu,", value );
+}
+
+static void
+jsonp_long( fd_gui_t *   gui,
+            char const * key,
+            long         value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%lu,", key, value );
+  else                   fd_http_server_printf( gui->http, "%ld,", value );
+}
+
+static void 
+jsonp_double( fd_gui_t *   gui,
+              char const * key,
+              double       value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%.2f,", key, value );
+  else                   fd_http_server_printf( gui->http, "%.2f,", value );
+}
+
+static void
+jsonp_sanitize_str( fd_http_server_t * http,
+                    ulong              start_len ) {
+  /* escape quotemark, reverse solidus, and control chars U+0000 through U+001F
+     just replace with a space */
+  uchar * data = http->oring;
+  for( ulong i=start_len; i<http->stage_len; i++ ) {
+    if( FD_UNLIKELY( data[ (http->stage_off%http->oring_sz)+i ] < 0x20 ||
+                     data[ (http->stage_off%http->oring_sz)+i ] == '"' ||
+                     data[ (http->stage_off%http->oring_sz)+i ] == '\\' ) ) {
+      data[ (http->stage_off%http->oring_sz)+i ] = ' ';
+    }
+  }
+}
+
+static void
+jsonp_string( fd_gui_t *   gui,
+              char const * key,
+              char const * value ) {
+  char * val = (void *)value;
+  if( FD_LIKELY( value ) ) {
+    if( FD_UNLIKELY( !fd_utf8_verify( value, strlen( value ) ) )) {
+      val = NULL;
+    }
+  }
+  if( FD_LIKELY( key ) ) {
+    if( FD_LIKELY( val ) ) {
+      fd_http_server_printf( gui->http, "\"%s\":\"", key );
+      ulong start_len = gui->http->stage_len;
+      fd_http_server_printf( gui->http, "%s", val );
+      jsonp_sanitize_str( gui->http, start_len );
+      fd_http_server_printf( gui->http, "\"," );
+    } else {
+      fd_http_server_printf( gui->http, "\"%s\":null,", key );
+    }
+  } else {
+    if( FD_LIKELY( val ) ) {
+      fd_http_server_printf( gui->http, "\"" );
+      ulong start_len = gui->http->stage_len;
+      fd_http_server_printf( gui->http, "%s", val );
+      jsonp_sanitize_str( gui->http, start_len );
+      fd_http_server_printf( gui->http, "\"," );
+    } else {
+      fd_http_server_printf( gui->http, "null," );
+    }
+  }
```

Comment:
What do you think about this? One extra call to printf in the good case, but less repetition.
```suggestion
  if( FD_LIKELY( key ) ) {
    fd_http_server_printf( gui->http, "\"%s\":", key );
  }
  if( FD_LIKELY( val ) ) {
    fd_http_server_printf( gui->http, "\"", key );
    ulong start_len = gui->http->stage_len;
    fd_http_server_printf( gui->http, "%s", val );
    jsonp_sanitize_str( gui->http, start_len );
    fd_http_server_printf( gui->http, "\"," );
  } else {
    fd_http_server_printf( gui->http, "null," );
  }
 ```